### DISCUSSION TITLE
[Design System, ui] Vertical tabs or Menus?

### DISCUSSION THREAD

**Discussion by cooper-joe** [dhis2/notes#194] [Ideas]
# Background
[There is a proposal to add a vertical variant](https://jira.dhis2.org/browse/DHIS2-9878) to the `Tab` `ui` component. A discussion opened up that suggested using existing `MenuItem`/`Menu` components to achieve the functionality that vertical `Tab`s would provide.

Providing a vertical tab/sidebar solution is currently blocking features that are planned for 2.36 release.

## Why vertical tabs?

The proposal of vertical tabs was intended to provide the following functionality: 
- Display options in a side-aligned sidebar
- Display many options more effectively/accessible than horizontal tabs
- Provide the ability to nest tabs/options within other tabs. Two sets of horizontal tabs is not an effective interface.

An example of vertical tab usage that led to this proposal:
![image](https://user-images.githubusercontent.com/33054985/103882746-8b98b480-50dc-11eb-8549-17380c93dc5c.png)

## Using `MenuItem`s instead

This functionality could indeed be provided by the `MenuItem` and `Menu` component, but with some adjustments to allow for sidebar functionality. These adjustments are outlined in this document.

Currently, the `MenuItem` and `Menu` component is designed specifically with a `DropdownMenu` in mind. The [Menu component in the Design System](https://github.com/dhis2/design-system/blob/master/molecules/menu.md) only describes it's use as a dropdown menu. Using `MenuItem`s for a sidebar-type component is possible, but not fully supported. 

## Dropdown `MenuItem`s and Sidebar `MenuItem`s

The requirements of a `MenuItem` are similar, but not identical, for a sidebar and a dropdown menu:

**Sidebar menu items**
	- Enabled
	- Hover
	- Focus
	- Active (click)
	- Disabled
	- Selected
	- Expanded/Collapsed
	- Display icon
	- Dense option

**Dropdown menu items**
	- Enabled
	- Hover
	- Focus
	- Active (click)
	- Disabled
	- Toggled
	- Open (child menu open from this item)
	- Display icon
	- Dense option

A visual comparison of the states required by a `MenuItem` in these two different use cases:

![image](https://user-images.githubusercontent.com/33054985/103882638-686e0500-50dc-11eb-8c0a-c59f23522f7d.png)

(Styles are for illustration only and are not the final styles for sidebar menu items)

## Requirements of a dual-usage `MenuItem`
To support the dual-usage, a `MenuItem` needs to provide the following states, with clear differentiation between each:
- Enabled
- Disabled
- Hover
- Focus
- Active (as in, clicked)
- Toggled on / off (multiple allowed)
- Selected (only one allowed at any one time)

A `MenuItem` that satisfies the above requirements would then be composed for the different use cases, as follows:

Component: `DropdownMenu`
- composed using a `DropdownContainer` and multiple `MenuItem`s

Component: `SidebarMenu`
- composed using a `SidebarContainer` and multiple `MenuItem`s

## Work required
To use a `MenuItem` for the dual-purpose of a dropdown and sidebar type menu, the following changes are needed:
- Provide expand/collapse functionality and styles for `MenuItem`s with descendants in a `SidebarMenu`
- Provide a selected state for `MenuItem`s (used only in `SidebarMenu`)
- Provide a `SidebarMenu` container component

---

### Discussion

 - Reusing components is where possible is ideal, but are we trying to use a single component for two use cases that are too different? This could be the case, highlighted by the fact that there are certain states that are specific to use cases (`toggled` for dropdown, `selected` for sidebar).
 - Should in-page content switching be handled by `Tabs` only? A `SidebarMenu` makes sense for the main page layout, but should it be used inside a page too? 
- Are there other drawbacks to the above approach?

**Comment by HendrikThePendric** [comment]
I will provide a more detailed answer later, but for now, I just wanted to point out that we briefly discussed this very use case when working on the menu, and this the reason we ended up with both a `Menu` (with very little behaviour) and a `FlyoutMenu` (which deals with sub-menus in popovers etc.).

I need to think about this some more but I think in general terms we need to chose between two approaches:
A. Introduce new components for this vertical tab UI and remove some menu-related components to avoid overlap / confusion
B. Introduce more menu-related components / add functionality to existing ones
My initial preference would be B, but if this gets too messy then A would be OK with me too.

**Comment by ghost** [comment]
Same here. I'll also need some time to really be able to say something sensible. One thing that immediately comes to mind is the intended usecases for both the tabs and the menu (and especially the vertically oriented variants). What are they supposed to be used for? For me that's where I'd start with figuring out whether the components are actually different, and if they are if the differences are cosmetic or functional.

**Comment by cooper-joe** [comment]
Thanks for the initial thoughts, I can add some information regarding usecases:

> One thing that immediately comes to mind is the intended usecases for both the tabs and the menu (and especially the vertically oriented variants). What are they supposed to be used for?

The use cases for the components are:
- **Tabs**: navigate between related sections of content within a page, modal, section, container etc.
	- **Vertical tabs, in particular**: allow nesting of tabs within tabs in an understandable way, allow large number of tabs, 
- **Menu (Flyout / Dropdown)**: select, change or trigger options from a container that opens upon activating a trigger (such as a button).
- **Menu (Sidebar)**: navigate between sections of an application (potentially unrelated to one another, unlike tabs)

Based on the use cases above, I think there is a reasonable argument for "all of the above" being available in `ui`.

**Reply by ghost** [reply]
To me the usecases are a bit similar. A tab to me implies (as you also mention), that the 'navigation' is within a page. So that the content is already loaded, but just not displayed.

For the vertical tab, allowing a large number of tabs would not be ideal, as that would slow down loading of the page a lot (because you're loading a lot of content, of which you're only displaying a small portion). In that case, a menu seems more appropriate because loading the content separately is a lot faster.

So I do wonder why we need the vertical tabs, and why a vertically oriented menu would not do. What usecase can't be covered with the menu that we want to cover with the vertical tabs specifically (given that a large number of tabs is not something I'd recommend)?

edit: I think the one usecase I see is using vertical tabs for separating content on a single page into a _small_ amount of separate topics. If it's too much content I think it would benefit from being separated into separate routes, both for the user and for the loading speed. So I don't think I'd allow for nesting or a scrollable container. And vertical orientation is something we'd add then because I can imagine that there are usecases where that just makes more sense for the design. That would make the change relatively simple, and also keep the usecases clear (in my opinion).

**Reply by ghost** [reply]
Oh by the way, we should keep the discussion threaded. Assuming that your reply here is to my question above @cooper-joe, you can reply under my initial question comment and it'll keep it contained in a single thread.

**Reply by cooper-joe** [reply]
> So I do wonder why we need the vertical tabs, and why a vertically oriented menu would not do. What usecase can't be covered with the menu that we want to cover with the vertical tabs specifically (given that a large number of tabs is not something I'd recommend)?

I think we can identify this, and potentially make the use cases less problematic, by emphasizing the differences.

Focusing on the difference between vertical tabs and a vertical sidebar menu:
**Vertical tabs**: navigation between **related content sections** within another component, page or section.
**Sidebar menu**: navigation between **distinct, potentially unrelated pages of an application**.

The key difference is what navigation possibilities these components offer. Tabs are for related content, Sidebar is the "main" navigation that can link unrelated pages of an application. 

It may well be the case that a single component (tabs or menu) can fulfill both of those requirements. I'm still working through this myself, I'm unsure what I think at the moment.

Here is a visual example of both of these components could exist on the same page, fulfilling different roles. I've given the sidebar a different (dark) style to emphasize the difference:
![image](https://user-images.githubusercontent.com/33054985/103901205-1e474c80-50f9-11eb-93ae-74aaed8c65cc.png)


Looking at the visual example, it's clear that both the sidebar and the vertical tabs are providing similar functionality: navigating between things. The question is: is there enough difference to warrant two components? 

---

As an aside, I think breaking these use cases into separate components also allows the components room to grow into the use case as those use cases develop. For example, a Sidebar Menu may one day need specific sidebar functionality like badge counts (for notifications), separate top/bottom sections, collapse/expand functionality, different color schemes like in the example above, and so on. Such functionality would likely not be needed in vertical tabs, so it seems at odds to combine them now.

**Reply by ghost** [reply]
Sorry for the late reply. I've thought about it a bit, hopefully I can clarify my thoughts on this a little.

> Vertical tabs: navigation between related content sections within another component, page or section.
> Sidebar menu: navigation between distinct, potentially unrelated pages of an application.

That definition sounds good to me. Though to fit my idea of tabs and a menu more precisely, I would change it to:

> Tabs: navigation between related content, without routing
> Sidebar menu: navigation between unrelated content, by routing

My reasoning behind that change, is that from a technical perspective to me there's one important distinction between tabs and a menu:

* Tabs accept content, and then _hide/show_ content based on which tab is active (so no navigation is occuring, the content is already loaded)
* A menu will usually have a handler, and _route_ to content (so the content isn't pre-loaded)

Because tabs and menus look a bit similar. So what is the distinction between them? In my opinion it's that tabs exist to ease sectioning related content. By preloading content with the tabs, they can be switched to easily and quickly, and the transistion is not as jarring as routing. The downside being that you're preloading the content, so you can't use it to load a _lot_ of content, because that'd slow down that section of the app. In that case a menu is more appropriate. This fits with the explanation here: https://ux.stackexchange.com/a/23528 (the way I read it), so hopefully my assumptions aren't too idiosyncratic.

That means that these two pieces of functionality from the top comment conflict with my definition: 

* Display _many_ options more effectively/accessible than horizontal tabs
* Provide the ability to _nest_ tabs/options within other tabs. Two sets of horizontal tabs is not an effective interface.

The reason that they conflict is that, if you accept my definition, you don't want to cater to the usecase of sectioning off _a lot_ of content (because of the slowing down).

I hope that helps. Note that I think that my idea of the difference between tabs and menu does not fit with the current reality of @dhis2/ui. I believe tabs don't currently work as I've defined above. But since we're hitting a problem where the distinction between both isn't clear, I'd tend towards the definitions above to solve that.

**Reply by cooper-joe** [reply]
I agree @ismay, performance is an important part of user-experience and, as you said, loading a lot of content that is not visible is potentially problematic.

I think we can move forward with the simple defenition you laid out:
>Tabs: navigation between related content, without routing
>Sidebar menu: navigation between unrelated content, by routing
I think the difference between these two use cases is clear enough that having two components, `Tabs` and `Menu`, is justified.

After resolving these definitions, the next discussion from this point is the differences highlighted in the original post between `DropdownMenu` and `SidebarMenu`.

**Reply by ghost** [reply]
That sounds good to me 👍

> I think the difference between these two use cases is clear enough that having two components, Tabs and Menu, is justified.

I agree 👍.
